storage/store/collections.go

```diff
@@ -0,0 +1,173 @@
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve collection: %w", err)
+	}
+
+	for _, txID := range light.Transactions {
+		tx, err := c.transactions.ByID(txID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve transaction: %w", err)
+		}
+
+		collection.Transactions = append(collection.Transactions, tx)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &collection, nil
+}
+
+func (c *Collections) LightByID(colID flow.Identifier) (*flow.LightCollection, error) {
+	var collection flow.LightCollection
+
+	err := operation.RetrieveCollection(c.db.Reader(), colID, &collection)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve collection: %w", err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &collection, nil
+}
+
+func (c *Collections) Remove(colID flow.Identifier) error {
+	err := c.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
+		return operation.RemoveCollection(rw.Writer(), colID)
+	})
+	if err != nil {
+		return fmt.Errorf("could not remove collection: %w", err)
+	}
+	return nil
+}
+
+func (c *Collections) StoreLightAndIndexByTransaction(collection *flow.LightCollection) error {
+	// - This lock is to ensure there is no race condition when indexing collection by transaction ID
+	// - The access node uses this index to report the transaction status. It's done by first
+	//   find the collection for a given transaction ID, and then find the block by the collection,
+	//   and then find the status of the block.
+	// - since a transaction can belong to multiple collections, when indexing collection by transaction ID,
+	//   if we overwrite the previous collection ID that was indexed by the same transaction ID, the access node
+	//   will return different collection for the same transaction, and the transaction result status will be
+	//   inconsistent.
+	// - therefore, we need to check if the transaction is already indexed by a collection, and to
+	//   make sure there is no dirty read, we need to use a lock to protect the indexing operation.
+	// - Note, this approach works because this is the only place where UnsafeIndexCollectionByTransaction
+	//   is used in the code base to index collection by transaction.
+	c.indexingByTx.Lock()
+	defer c.indexingByTx.Unlock()
+
+	return c.db.WithReaderBatchWriter(func(rw storage.ReaderBatchWriter) error {
+		err := operation.InsertCollection(rw.Writer(), collection)
+		if err != nil {
+			return fmt.Errorf("could not insert collection: %w", err)
+		}
+
+		for _, txID := range collection.Transactions {
+			var differentColTxIsIn flow.Identifier
+			err := operation.RetrieveCollectionID(rw.GlobalReader(), txID, &differentColTxIsIn)
+			if err == nil {
+				// transaction is already indexed by a different collection, we should not index it again
+				// so that the access node will always return the same collection for a given transaction
+				// and return a consistent transaction result status.
+				continue
```

Comment:
I think we should return an error here since LNs are supposed to prevent a tx from 
1. appearing multiple times in the same collection
2. appearing in multiple collections